File: leetcode/0443-string-compression/Solution.java
class Solution {
    public int compress(char[] chars) {
        int write = 0;
        int read = 0;
        int n = chars.length;

        while (read < n) {
            char c = chars[read];
            int start = read;
            while (read < n && chars[read] == c) {
                read++;
            }
            int count = read - start;

            chars[write++] = c;

            if (count > 1) {
                for (char digit : String.valueOf(count).toCharArray()) {
                    chars[write++] = digit;
                }
            }
        }

        return write;
    }
}

File: leetcode/0443-string-compression/solution.py
from typing import List


class Solution:
    def compress(self, chars: List[str]) -> int:
        write = 0
        read = 0
        n = len(chars)

        while read < n:
            char = chars[read]
            start = read
            while read < n and chars[read] == char:
                read += 1
            count = read - start

            chars[write] = char
            write += 1

            if count > 1:
                for digit in str(count):
                    chars[write] = digit
                    write += 1

        return write
